### PR TITLE
TriBITS: Change `yaml-cpp` to `yamlcpp`

### PR DESCRIPTION
**Pull request by cwschilly**
@trilinos/tribits 
@rppawlo 
@bartlettroscoe 

FYI @stmcgovern 

## Motivation

This PR renames the `yaml-cpp` TPL to `yamlcpp` so that the TriBITS-generated macro (`HAVE_TEUCHOSPARAMETERLIST_YAMLCPP`) will be a valid C/C++ macro.

This is in response to [ this comment](https://github.com/trilinos/Trilinos/pull/12599#discussion_r1445010890) on PR #12599, which added `yaml-cpp` as an optional TPL.

## Testing

This PR has been tested locally, but currently there is not a yamlcpp-enabled pipeline for the autotester.

### CODE REVIEW

**Review by bartlettroscoe — CHANGES_REQUESTED** [review]
@cwschilly, thanks for putting this this PR.  There are just a few changes suggested below to fully take advantage of this name change and to provide the best TriBITS compatibility.

File: packages/krino/krino/krino_lib/CMakeLists.txt
Comment (cwschilly) [reply packages/krino/krino/krino_lib/CMakeLists.txt]: Fixed

File: packages/kokkos-kernels/perf_test/performance/CMakeLists.txt
Comment (cwschilly) [reply packages/kokkos-kernels/perf_test/performance/CMakeLists.txt]: Fixed; what is the process to copy back to the KokkosKernels repo?

File: packages/teuchos/parameterlist/CMakeLists.txt
Comment (cwschilly) [reply packages/teuchos/parameterlist/CMakeLists.txt]: Fixed and tested locally

File: packages/kokkos-kernels/perf_test/performance/CMakeLists.txt
Comment (bartlettroscoe) [reply packages/kokkos-kernels/perf_test/performance/CMakeLists.txt]: The process involves pulling the commit off the Trilinos repo branch using `git format-patch` and then applying it to a Kokkos repo branch take off of the correct base commit using `git am`.  I wish this process was documented somewhere but I don't believe that it is.

**Review by bartlettroscoe — DISMISSED** [review]
@cwschilly, thanks for addressing those issues.  If the PR tests all pass, this is ready to merge.  Thanks!

**Comment by stmcgovern** [comment]
Hi @bartlettroscoe, we just spoke with @rppawlo and I believe he suggested holding off on this until the Kokkos promotion is integrated in (to avoid any possible KK snapshotting overwriting the work here). To be on the safe side, would you mind not merging this (removing the automerge label) yet?

**Review by ndellingwood — CHANGES_REQUESTED** [review]
Changes to kokkos-kernels (or kokkos) need a PR to the standalone repo approved by devs

**Comment by ndellingwood** [comment]
> there is a kokkos 4.2.1 release in a separate PR that might collide with these changes.

@rppawlo correct, and if this PR merges before #12707 then the changes here would be clobbered by the release PR.
Please hold off until #12707 is merged as well as separate PR to kokkos-kernels has been submitted and approved by devs

**Comment by bartlettroscoe** [comment]
> > there is a kokkos 4.2.1 release in a separate PR that might collide with these changes.
> 
> @rppawlo correct, and if this PR merges before #12707 then the changes here would be clobbered by the release PR. Please hold off until #12707 is merged as well as separate PR to kokkos-kernels has been submitted and approved by devs

@rppawlo and @ndellingwood, since KokkosKernels was already snapshotted into a version of the Trilinos 'develop' branch for #12707, there should be a smooth merge between this PR #12701 and PR #12707.  That is, there should be no reason to hold off on merging these PRs to Trilinos 'develop' in either order.   Since `yamlcpp`/`yaml-cpp` is not enabled in PR testing yet, there would not be any configure or build errors even if there was a semantic conflict.   If I am mistaken, please let me know.

However, the **next time** KokkosKernels in snapshotted, we need to make sure that all of the Trilinos-side changes are in the KokkosKernels repo branch for the source.

FYI: The TriBITS approach to snapshotting ensures that overwriting does not occur between TriBITS snapshots and local changes to TriBITS in Trilinos so we never worry about issues like this with TriBITS.

**Comment by bartlettroscoe** [comment]
> The kokkos-kernels changes in this PR still need to be submitted and approved at the standalone kokkos-kernels repo before they should merge to Trilinos

@ndellingwood, I will pull off the KokkosKernels commit(s) from this topic branch and create the kokkos-kernels PR.  (If I have trouble determining the base kokkos-kernels snaphshot version, I will let you know.)

File: packages/krino/krino/krino_lib/CMakeLists.txt
Comment (bartlettroscoe) [inline packages/krino/krino/krino_lib/CMakeLists.txt:8]: That is the ticket 👍

**Review by bartlettroscoe — APPROVED** [review]
That should fix the Krino build errors.  (I should have noticed that in my last review.)

**Comment by bartlettroscoe** [comment]
> The kokkos-kernels changes in this PR still need to be submitted and approved at the standalone kokkos-kernels repo before they should merge to Trilinos

@ndellingwood, I created the kokkos-kernels PR:

* kokkos/kokkos-kernels#2099

that matches the changes to KokkosKernels in this PR. 

This was a straightforward process, except for an expected merge conflict due to changes in the `kokkos-kerenls/cmake/Dependencies.cmake` file because of kokkos-kernels 'develop' commits kokkos/kokkos-kernels@4f3549d995e4883c4cb05743850236ec71be2e88, kokkos/kokkos-kernels@18ca910d3f13fbb977fed98f88e84dbb5d4d776d, kokkos/kokkos-kernels@24c73c8f73ec3ac0c663b61cb2130ca6908d4cbe that touch those same lines that are not reflected in the Trilinos 'develop' version of `packages/kokkos-kernels`.  So the line was a complete conflict and the rename had to be done manually (but it was obvious and trivial).

**Comment by bartlettroscoe** [comment]
FYI: The experience in creating:

* https://github.com/kokkos/kokkos-kernels/pull/2099

shows that there will be a merge conflict between this PR topic branch and that for #12707 in the `kokkos-kernels/cmake/Dependencies.cmake` file.  So whichever PR merges first will block the merge of the other PR until someone resolves the merge conflict in the latter PR.  (The problem is not related to the snapshotting process.  The problem is conflicting changes made to that file in the Trilinos 'develop' vs. kokkos-kernels 'develop' branches that no version control system that I know of can currently resolve automatically.)

So we will have to wait for the merge of PR:

* #12707

and then have to resolve the conflict in this PR before it can be merged ☹️

**Review by ndellingwood — APPROVED** [review]
Patch match to kokkos-kernels submitted https://github.com/kokkos/kokkos-kernels/pull/2099 , thanks @bartlettroscoe !

**Comment by bartlettroscoe** [comment]
@ndellingwood, okay to merge this PR now?

**Comment by ndellingwood** [comment]
> @ndellingwood, okay to merge this PR now?

Yes, I'll click the button 👍

**Pull request merged**
